contracts/proposal/cwd-proposal-delegate/src/contract.rs

```diff
@@ -0,0 +1,238 @@
+        }
+        ExecuteMsg::Execute { delegation_id } => execute_execute(deps, env, info, delegation_id),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ContractError> {
+    match msg.id {
+        REPLY_ID_EXECUTE_PROPOSAL_HOOK => {
+            let id = EXECUTE_CTX.load(deps.storage)?;
+            let Delegation {
+                policy_preserve_on_failure,
+                ..
+            } = DELEGATIONS.load(deps.storage, id)?;
+
+            if let SubMsgResult::Err(_) = msg.result {
+                // && with `let` expression above gives unstable Rust error
+                if policy_preserve_on_failure {
+                    return Ok(Response::default()
+                        .add_attribute("execute_failed_but_preserved", id.to_string()));
+                }
+            }
+            // Delete delegation in both success and error case
+            DELEGATIONS.remove(deps.storage, id);
+            Ok(Response::default())
+        }
+        _ => Err(ContractError::Std(StdError::GenericErr {
+            msg: "Reply handler ID not found".into(),
+        })),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::DelegationCount {} => Ok(to_binary(&query_delegation_count(deps)?)?),
+        QueryMsg::Delegation { delegation_id } => {
+            Ok(to_binary(&query_delegation(deps, delegation_id)?)?)
+        }
+    }
+}
+
+// MARK: Query helpers
+
+fn query_delegation_count(deps: Deps) -> StdResult<DelegationCountResponse> {
+    let count = DELEGATION_COUNT.load(deps.storage)?;
+    Ok(DelegationCountResponse { count })
+}
+
+fn query_delegation(deps: Deps, delegation_id: u64) -> StdResult<DelegationResponse> {
+    let delegation = DELEGATIONS.load(deps.storage, delegation_id)?;
+    Ok(delegation)
+}
+
+// MARK: Execute subroutines
+
+fn execute_delegate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    delegation: Delegation,
+) -> Result<Response, ContractError> {
+    let Config { admin } = CONFIG.load(deps.storage)?;
+    if info.sender != admin {
+        return Err(ContractError::Unauthorized {});
+    }
+    assert_not_expired(&delegation.expiration, &env.block)?;
+
+    let id = advance_delegation_count(deps.storage)?;
+    DELEGATIONS.save(deps.storage, id, &delegation)?;
+
+    Ok(Response::default().add_attribute("delegate_id", id.to_string()))
+}
+
+fn execute_remove_delegation(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    delegation_id: u64,
+) -> Result<Response, ContractError> {
+    let Config { admin } = CONFIG.load(deps.storage)?;
+    if info.sender != admin {
+        return Err(ContractError::Unauthorized {});
+    }
+    // If delegation is irrevocable, return Error
+    let Delegation {
+        policy_irrevocable, ..
```

Comment:
what's the reason for having this option? i feel things like this can give fake guarantees if the contract is instantiated with a DAO as it's admin. should we be instantiating this without an admin perhap?

Reply:
good point. The DAO could always remove the proposal module and all delegations would be revocable. However, that would eliminate all their existing delegations as well. There's 2 options to go with this:
1. Rename the policy to `policy_proposal_irrevocable` and add a warning
2. Remove the policy altogether